packages/react-router/modules/__tests__/resolve-test.js

```diff
@@ -0,0 +1,159 @@
+import expect from 'expect'
+import { resolveLocation, resolvePath } from '../resolve'
+
+describe('resolve', () => {
+  describe('resolvePath', () => {
+    it('returns pathname if it is absolute', () => {
+      const pathname = '/recipes'
+      expect(resolvePath(pathname, '/base')).toBe('/recipes')
+    })
+
+    it('joins pathname with the base', () => {
+      const pathname = 'recipes'
+      expect(resolvePath(pathname, '/base')).toBe('/base/recipes')
+    })
+
+    it('correctly joins when base has a trailing slash', () => {
+      const pathname = 'recipes'
+      expect(resolvePath(pathname, '/base/')).toBe('/base/recipes')
+    })
+
+    it('returns undefined when pathname is undefined', () => {
+      const pathname = undefined
+      expect(resolvePath(pathname, '/base')).toBe(undefined)
+    })
+
+    describe('empty base', () => {
+      const cases = [
+        ['', '/'],
+        ['recipes', '/recipes'],
+        ['/restaurants', '/restaurants'],
+        [undefined, undefined]
+      ]
+
+      it('works when base is empty string', () => {
+        const BASE = ''
+        cases.forEach(([input, output]) => {
+          expect(resolvePath(input, BASE)).toEqual(output)
+        })
+      })
+
+      it('works when base is null', () => {
+        const BASE = null
+        cases.forEach(([input, output]) => {
+          expect(resolvePath(input, BASE)).toEqual(output)
+        })
+      })
+
+      it('works when base is undefined', () => {
+        const BASE = undefined
+        cases.forEach(([input, output]) => {
+          expect(resolvePath(input, BASE)).toEqual(output)
+        })
+      })
+    })
+  })
+
+  describe('resolveLocation', () => {
```

Comment:
There is no exported `resolve` function. The test name was just me grouping related tests, but I dropped that in the last commit.

I think that it would make more sense to merge `resolvePath` into the `matchPath` function because both `Route` and `Switch` call it. I'll push a commit later that does that for you to take a look at.